file/sequence_file_reader.cc

```diff
@@ -48,53 +62,87 @@ IOStatus SequentialFileReader::Read(size_t n, Slice* result, char* scratch) {
     buf.Alignment(alignment);
     buf.AllocateNewBuffer(size);
 
-    Slice tmp;
-    uint64_t orig_offset = 0;
-    FileOperationInfo::StartTimePoint start_ts;
-    if (ShouldNotifyListeners()) {
-      orig_offset = aligned_offset + buf.CurrentSize();
-      start_ts = FileOperationInfo::StartNow();
+    while (buf.CurrentSize() < size) {
+      size_t allowed;
+      if (rate_limiter_priority != Env::IO_TOTAL && rate_limiter_ != nullptr) {
+        allowed = rate_limiter_->RequestToken(
+            buf.Capacity() - buf.CurrentSize(), buf.Alignment(),
+            rate_limiter_priority, nullptr, RateLimiter::OpType::kRead);
+      } else {
+        assert(buf.CurrentSize() == 0);
+        allowed = size;
+      }
+
+      Slice tmp;
+      uint64_t orig_offset = 0;
+      FileOperationInfo::StartTimePoint start_ts;
+      if (ShouldNotifyListeners()) {
+        orig_offset = aligned_offset + buf.CurrentSize();
+        start_ts = FileOperationInfo::StartNow();
+      }
+      io_s =
+          file_->PositionedRead(aligned_offset + buf.CurrentSize(), allowed,
+                                IOOptions(), &tmp, buf.Destination(), nullptr);
+      if (ShouldNotifyListeners()) {
+        auto finish_ts = FileOperationInfo::FinishNow();
+        NotifyOnFileReadFinish(orig_offset, tmp.size(), start_ts, finish_ts,
+                               io_s);
+      }
+      buf.Size(buf.CurrentSize() + tmp.size());
+      if (!io_s.ok() || tmp.size() < allowed) {
+        break;
+      }
     }
-    io_s = file_->PositionedRead(aligned_offset, size, IOOptions(), &tmp,
-                                 buf.BufferStart(), nullptr);
-    if (io_s.ok() && offset_advance < tmp.size()) {
-      buf.Size(tmp.size());
+
+    if (io_s.ok() && offset_advance < buf.CurrentSize()) {
       r = buf.Read(scratch, offset_advance,
-                   std::min(tmp.size() - offset_advance, n));
+                   std::min(buf.CurrentSize() - offset_advance, n));
```

Comment:
Not relevant to @cbi42 's impl but just for my learning:

cc @anand1976 our file read wrapper expert - not relevant to this PR's impl but it seems to me that 
(1) We are not returning the first `offset_advance` bytes in the `AlignedBuf buf` by `r = buf.Read(scratch, offset_advance,...)` to user's result. Are those bytes garbage? I understand that we need to read aligned but don't understand why we abandon those bytes.

(2) It also seems to me that `offset_advance < buf.CurrentSize()` is trying to guard from the **only** edge case where `offset_advance == buf.CurrentSize()`, which is the case **only where** `n==0` and offset is right at some alignment (i.e, `offset % alignment = 0`). This further implies `offset_advance == 0` and `buf.CurrentSize() == 0`.

Am I understanding this correctly?

Reply:
From my understanding:
(1) The first `offset_advance` bytes are the bytes from `aligned_offset` to `offset`, which are probably already read by the caller in previous calls, and the caller is expecting to read sequentially for bytes after `offset`, so these `offset_advance` bytes before `offset` are discarded.
(2) I think it is possible to have `offset_advance > buf.CurrentSize()`, e.g., when the call to `file_->PositionedRead` returns ok status but `tmp.size() < offset_advance`. Not sure if that's likely tho.
If we consider pre-PR when file size is not considered, then if `Read` is called when `offset` and `aligned_offset` are already at end of file, then I think we could have offset_advance > buf.CurrentSize() too.

Reply:
@cbi42 aha - thanks! Seems like we could use some clarification like this https://github.com/facebook/rocksdb/pull/10002 :)